src/compiler/checker.ts

```diff
@@ -24056,7 +24057,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function getTypeOfReverseMappedSymbol(symbol: ReverseMappedSymbol) {
         const links = getSymbolLinks(symbol);
         if (!links.type) {
+            // Since `inferReverseMappedType` avoids pulling on structure in its inference result, we *shouldn't* have a recursive
+            // type resolution request, but if we somehow do, we should error.
+            if (!pushTypeResolution(symbol, TypeSystemPropertyName.Type)) {
+                reportCircularityError(symbol);

```

Comment:
In the current version of this PR, this error is never reported, since I've avoided pulling on the structure elsewhere, but this is what's needed to prevent another stack depth error from happening with any other incidental recursion in reverse mapped type resolution. Plus, basically every other branch of `getTypeOfSymbol` has a guard like this. Cue "how did we even manage to get by without circularity guards here?" response.